### PR TITLE
Set custom per instance socket address

### PR DESCRIPTION
**Pull request by kurtisvg**
Fork of #265, updated to include automatic addition of the suffix for Postgres instances.

### CODE REVIEW

File: cmd/cloud_sql_proxy/proxy.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ import (
 	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/fuse"
 	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy"
 	"github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/util"
-	sqladmin "google.golang.org/api/sqladmin/v1beta4"
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: revert

File: cmd/cloud_sql_proxy/proxy.go [hunk 2]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
+		return instanceConfig{}, fmt.Errorf("invalid instance argument: must be either form - `<instance_connection_string>` or `<instance_connection_string>=<options>`; invalid arg was %q", instance)
+	}
+	// Parse the instance connection name - everything before the "=".
+	ret.Instance = args[0]
+	proj, _, name := util.SplitName(ret.Instance)
+	if proj == "" || name == "" {
+		return instanceConfig{}, fmt.Errorf("invalid instance connection string: must be in the form `project:region:instance-name`; invalid name was %q", args[0])
+	}
+	// Parse the instance options - everything before the "=".
+	eM := "invalid instance options: must be in the form `unix:/path/to/socket`, `tcp:port`, `tcp:host:port`; invalid option was %q"
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: what is eM? error message? please inline where it's used (L254)

File: cmd/cloud_sql_proxy/proxy.go [hunk 3]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
+	if len(args) == 2 {
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: a bit clearer to have the shorter code path first, i think.

if len(args) == 1 {
  // default unix
} else {
  // parse options
}

File: cmd/cloud_sql_proxy/proxy.go [hunk 4]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
+			}
+			ret.Network = "unix"
+			if strings.HasPrefix(opts[1], "/") {
+				ret.Address = opts[1] // Root path.
+			} else {
+				ret.Address = filepath.Join(dir, opts[1])
+			}
+		} else {
+			// Listen via the specified dial address.
+			ret.Network = opts[0]
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: move up to L237 (no need for L243)

File: cmd/cloud_sql_proxy/proxy.go [hunk 5]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
+			switch len(opts) {
+			default:
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: default at the bottom

File: cmd/cloud_sql_proxy/proxy.go [hunk 6]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
+				return instanceConfig{}, fmt.Errorf(eM, args[1])
+			case 2:
+				// No "host" part of the address. Be safe and assume that they want a loopback address.
+				addr, ok := loopbackForNet[opts[0]]
+				if !ok {
+					return ret, fmt.Errorf("invalid %q: unrecognized network %v", args[1], opts[0])
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: s/unrecognized/unsupported/

File: cmd/cloud_sql_proxy/proxy.go [hunk 7]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
+				}
+				ret.Address = fmt.Sprintf("%s:%s", addr, opts[1])
+			case 3:
+				// User provided a host and port; use that.
+				ret.Address = fmt.Sprintf("%s:%s", opts[1], opts[2])
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: use net.JoinHostPort

https://golang.org/pkg/net/#JoinHostPort

File: cmd/cloud_sql_proxy/proxy.go (hunk 5 above)
Comment (broady) [reply cmd/cloud_sql_proxy/proxy.go]: also, unfortunate this doesn't support ipv6 addresses very well.

File: cmd/cloud_sql_proxy/proxy.go (hunk 2 above)
Comment (kurtisvg) [reply cmd/cloud_sql_proxy/proxy.go]: It's used on L241 and L254 - I would prefer to keep the same error for both lines

File: cmd/cloud_sql_proxy/proxy.go [hunk 8]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
-		if proj == "" || name == "" {
-			return instanceConfig{}, fmt.Errorf("invalid instance name: must be in the form `project:region:instance-name`; invalid name was %q", instance)
-		}
-		// We allow people to omit the region due to historical reasons. It'll
-		// fail later in the code if this isn't allowed, so just assume it's
-		// allowed until we actually need the region in this API call.
-		in, err := sql.Instances.Get(proj, name).Do()
-		if err != nil {
+	// Use the SQL Admin API to verify compatibility with the instance.
+	sql, err := sqladmin.New(cl)
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: very surprising for HTTP requests to happen inside a "parseX" function. split them out?

File: cmd/cloud_sql_proxy/proxy.go [hunk 9]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
+	if err != nil {
+		return instanceConfig{}, err
+	}
+	sql.BasePath = *host
+	inst, err := sql.Instances.Get(proj, name).Do()
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: probably want some retries here

File: cmd/cloud_sql_proxy/proxy.go [hunk 10]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
+	if err != nil {
+		return instanceConfig{}, err
+	}
+	if inst.BackendType == "FIRST_GEN" {
+		logging.Errorf("WARNING: proxy client does not support first generation Cloud SQL instances.")
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: doesn't look like a warning. but i notice there's the word "WARNING" everywhere even though things are not warnings but actual fatal errors.

File: cmd/cloud_sql_proxy/proxy.go [hunk 11]
```diff
@@ -221,62 +221,77 @@ var validNets = func() map[string]bool {
...
+		return instanceConfig{}, fmt.Errorf("%q is a first generation instance", instance)
+	}
+	// Postgres instances use a special suffix on the unix socket.
+	// See https://www.postgresql.org/docs/11/runtime-config-connection.html
+	if ret.Network == "unix" && strings.HasPrefix(strings.ToLower(inst.DatabaseVersion), "postgres") {
+		// Verify the folder exists.
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: s/folder/directory/

File: cmd/cloud_sql_proxy/proxy.go (hunk 2 above)
Comment (broady) [reply cmd/cloud_sql_proxy/proxy.go]: ok, please give it a better name or a concrete type, like:

```
type fooError {
  invalidOption string
}

func (e fooError) Error() string {
  return fmt.Sprintf(...)
}
```

File: cmd/cloud_sql_proxy/proxy.go (hunk 6 above)
Comment (kurtisvg) [reply cmd/cloud_sql_proxy/proxy.go]: leaving: existing behavior

File: cmd/cloud_sql_proxy/proxy.go (hunk 8 above)
Comment (kurtisvg) [reply cmd/cloud_sql_proxy/proxy.go]: leaving: existing behavior

File: cmd/cloud_sql_proxy/proxy.go (hunk 9 above)
Comment (kurtisvg) [reply cmd/cloud_sql_proxy/proxy.go]: leaving: existing behavior

File: cmd/cloud_sql_proxy/proxy.go (hunk 10 above)
Comment (kurtisvg) [reply cmd/cloud_sql_proxy/proxy.go]: leaving: existing behavior

**Comment by wedi** [comment]
Thanks for stepping in. Ofc I'm fine with that. 👍

File: cmd/cloud_sql_proxy/proxy.go [hunk 12]
```diff
@@ -285,6 +280,33 @@ func parseInstanceConfig(dir, instance string, cl *http.Client) (instanceConfig,
...
+		return "", fmt.Errorf("invalid instance options: must be in the form `unix:/path/to/socket`, `tcp:port`, `tcp:host:port`; invalid option was %q", strings.Join(opts, ":"))
+	}
+	if strings.HasPrefix(opts[1], "/") {
+		return opts[1], nil // Root path.
+	}
+	return filepath.Join(dir, opts[1]), nil
+}
+
+// parseTcpOpts parses the instance options when specifying tcp port options.
+func parseTcpOpts(opts []string) (string, error) {
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: `(net, addrOpt string)`

File: cmd/cloud_sql_proxy/proxy.go [hunk 13]
```diff
@@ -221,62 +221,57 @@ var validNets = func() map[string]bool {
...
+		return instanceConfig{}, fmt.Errorf("invalid instance connection string: must be in the form `project:region:instance-name`; invalid name was %q", args[0])
+	}
+	if len(args) == 1 {
+		// Default to listening via unix socket in specified directory
+		ret.Network = "unix"
+		ret.Address = filepath.Join(dir, instance)
 	} else {
-		sql, err := sqladmin.New(cl)
+		// Parse the instance options if present.
+		opts := strings.SplitN(args[1], ":", 2)
```
Comment (broady) [inline cmd/cloud_sql_proxy/proxy.go]: bounds check on opts (len == 2)

**Pull request merged**